File: ios/nodes/OscillatorNode/OscillatorNode.m
#import <OscillatorNode.h>
#import "Constants.h"

@implementation OscillatorNode

- (instancetype)initWithContext:(AudioContext *)context
{
    if (self = [super initWithContext:context]) {
        
        _frequencyParam = [[AudioParam alloc] initWithContext:context
                                                        value:440
                                                     minValue:-[Constants nyquistFrequency]
                                                     maxValue:[Constants nyquistFrequency]];
        _detuneParam = [[AudioParam alloc] initWithContext:context
                                                     value:0
                                                  minValue:-[Constants maxDetune]
                                                  maxValue:[Constants maxDetune]];
        _waveType = WaveTypeSine;
        _isPlaying = NO;
        
        self.numberOfOutputs = 1;
        self.numberOfInputs = 0;
        
        _format = [[AVAudioFormat alloc] initStandardFormatWithSampleRate:self.context.sampleRate channels:2];
        _phase = 0.0;
        
        __weak typeof(self) weakSelf = self;
        _sourceNode = [[AVAudioSourceNode alloc] initWithFormat:_format
                                                    renderBlock:^OSStatus(
                                                                          BOOL *isSilence,
                                                                          const AudioTimeStamp *timestamp,
                                                                          AVAudioFrameCount frameCount,
                                                                          AudioBufferList *outputData) {
                                                                              return [weakSelf renderCallbackWithIsSilence:isSilence
                                                                                                                 timestamp:timestamp
                                                                                                                frameCount:frameCount
                                                                                                                outputData:outputData];
                                                                          }];
        
        [self.context.audioEngine attachNode:self.sourceNode];
        [self.context.audioEngine connect:self.sourceNode to:self.context.audioEngine.mainMixerNode format:self.format];
        
        if (!self.context.audioEngine.isRunning) {
            NSError *error = nil;
            if (![self.context.audioEngine startAndReturnError:&error]) {
                NSLog(@"Error starting audio engine: %@", [error localizedDescription]);
            }
        }
        
    }

  return self;
}

- (void)cleanup
{
  if (_isPlaying) {
    [self stopPlayback];
  }

  [self.context.audioEngine detachNode:self.sourceNode];

  _frequencyParam = nil;
  _detuneParam = nil;
  _format = nil;
}

- (void)start:(double)time
{
  if (_isPlaying) {
    return;
  }

  double delay = time - [self.context getCurrentTime];
  if (delay <= 0) {
    [self startPlayback];
  } else {
    [NSTimer scheduledTimerWithTimeInterval:delay
                                     target:self
                                   selector:@selector(startPlayback)
                                   userInfo:nil
                                    repeats:NO];
  }
}

- (void)startPlayback
{
  _isPlaying = YES;
}

- (void)stop:(double)time
{
  if (!_isPlaying) {
    return;
  }

  double delay = time - [self.context getCurrentTime];
  if (delay <= 0) {
    [self stopPlayback];
  } else {
    [NSTimer scheduledTimerWithTimeInterval:delay target:self selector:@selector(stopPlayback) userInfo:nil repeats:NO];
  }
}

- (void)stopPlayback
{
  _isPlaying = NO;
}

- (OSStatus)renderCallbackWithIsSilence:(BOOL *)isSilence
                              timestamp:(const AudioTimeStamp *)timestamp
                             frameCount:(AVAudioFrameCount)frameCount
                             outputData:(AudioBufferList *)outputData
{
  if (outputData->mNumberBuffers < 2) {
    return noErr; // Ensure we have stereo output
  }

  float *leftBuffer = (float *)outputData->mBuffers[0].mData;
  float *rightBuffer = (float *)outputData->mBuffers[1].mData;

  float time = [self.context getCurrentTime];
  float deltaTime = 1 / self.context.sampleRate;

  for (int frame = 0; frame < frameCount; frame += 1) {
    // Convert cents to HZ
    if (!_isPlaying) {
      leftBuffer[frame] = 0;
      rightBuffer[frame] = 0;
      continue;
    }

    double detuneRatio = pow(2.0, [_detuneParam getValueAtTime:time] / OCTAVE_IN_CENTS);
    double detunedFrequency = round(detuneRatio * [_frequencyParam getValueAtTime:time]);
    double phaseIncrement = FULL_CIRCLE_RADIANS * (detunedFrequency / self.context.sampleRate);
    leftBuffer[frame] = [WaveType valueForWaveType:_waveType atPhase:_phase];
    rightBuffer[frame] = [WaveType valueForWaveType:_waveType atPhase:_phase];

    _phase += phaseIncrement;
    time += deltaTime;

    if (_phase > FULL_CIRCLE_RADIANS) {
      _phase -= FULL_CIRCLE_RADIANS;
    }
  }

  [self process:frameCount bufferList:outputData];

  return noErr;
}

- (Boolean)getIsPlaying
{
  return _isPlaying;
}

- (void)setType:(NSString *)type
{
  _waveType = [WaveType waveTypeFromString:type];
}

- (NSString *)getType
{
  return [WaveType stringFromWaveType:self.waveType];
}

@end

File: packages/react-native-audio-api/common/cpp/core/OscillatorNode.cpp
#include "AudioBus.h"
#include "AudioArray.h"
#include "OscillatorNode.h"
#include "BaseAudioContext.h"

namespace audioapi {

OscillatorNode::OscillatorNode(BaseAudioContext *context)
    : AudioScheduledSourceNode(context) {
  frequencyParam_ = std::make_shared<AudioParam>(
      context, 444.0, -NYQUIST_FREQUENCY, NYQUIST_FREQUENCY);
  detuneParam_ =
      std::make_shared<AudioParam>(context, 0.0, -MAX_DETUNE, MAX_DETUNE);
  type_ = OscillatorType::SINE;
  periodicWave_ = context_->getBasicWaveForm(type_);
}

std::shared_ptr<AudioParam> OscillatorNode::getFrequencyParam() const {
  return frequencyParam_;
}

std::shared_ptr<AudioParam> OscillatorNode::getDetuneParam() const {
  return detuneParam_;
}

std::string OscillatorNode::getType() {
  return OscillatorNode::toString(type_);
}

void OscillatorNode::setType(const std::string &type) {
  type_ = OscillatorNode::fromString(type);
  periodicWave_ = context_->getBasicWaveForm(type_);
}

void OscillatorNode::setPeriodicWave(
    const std::shared_ptr<PeriodicWave> &periodicWave) {
  periodicWave_ = periodicWave;
  type_ = OscillatorType::CUSTOM;
}

void OscillatorNode::processNode(AudioBus* processingBus, int framesToProcess) {
  if (!isPlaying()) {
    processingBus->zero();
    return;
  }

  double time = context_->getCurrentTime();
  double deltaTime = 1.0 / context_->getSampleRate();

  for (int i = 0; i < framesToProcess; i += 1) {
    auto detuneRatio =
      std::pow(2.0f, detuneParam_->getValueAtTime(time) / 1200.0f);
    auto detunedFrequency =
        round(frequencyParam_->getValueAtTime(time) * detuneRatio);
    auto phaseIncrement = detunedFrequency * periodicWave_->getScale();

    float sample =
        periodicWave_->getSample(detunedFrequency, phase_, phaseIncrement);

    for (int j = 0; j < processingBus->getNumberOfChannels(); j += 1) {
      (*processingBus->getChannel(j))[i] = sample;
    }

    phase_ += phaseIncrement;
    phase_ -=
        floor(
            phase_ / static_cast<float>(periodicWave_->getPeriodicWaveSize())) *
        static_cast<float>(periodicWave_->getPeriodicWaveSize());

    time += deltaTime;
  }
}

} // namespace audioapi
